### PR TITLE
Cross entropy loss

### PR DESCRIPTION
**Pull request by evvahu**
This branch includes two methods to calculate loss from binary classification task and multiclass classification task. It further contains a test file with two tests.

- `multi_class_cross_entropy` calculates the negative loglikelihood and softmax in one go
- `binary_class_cross_entropy` applies a sigmoid function over the estimated levels and then applies cross-entropy over the value between 0 and 1.

other comments:
- for the multiclass loss: the target labels have to be of type _int_. At the moment I inserted an assert statement to make sure that this is the case. However, we might want to take it out or define it as a regex since `float32` does not account for all cases. Any suggestions?
- the tests were nice for me to actually understand what's going on. However, they're not really helpful since we can trust the library to implement the loss_functions the right way. Any suggestions for other tests?

### CODE REVIEW

File: tests/test_loss_functions.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    This test suite can be ran with:
+        python -m unittest -q tests.BasicTwoWordClassifierTest
+    """
+
+    def setUp(self):
+        self.output_tensor = torch.from_numpy(
+            np.array([[5.0, 5.0, 1.0], [0.0, 1.0, 2.0], [1.0, 2.0, 3.0]], dtype='float32'))
+        self.target_labels = torch.from_numpy(np.array([0, 1, 2], dtype='long'))
+        self.output_tensor_binary = torch.from_numpy(np.array([0.0, -1.4, -.8, .2, .4, .8, 1.2, 2.2, 2.9, 4.6]))
+        self.target_labels_binary = torch.from_numpy(np.array([1.0, 1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]))
```
Comment (Blubberli) [inline tests/test_loss_functions.py:19]: the targets are defined as floats and long (why long?) here. shouldn't it be int in both cases? or did you want to check whether your assertion gets raised?

File: tests/test_loss_functions.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+         """
+        loss = binary_class_cross_entropy(output=self.output_tensor_binary, target=self.target_labels_binary)
+
+        loss_calc = []
+        for x, y in zip(self.output_tensor_binary, self.target_labels_binary): # apply cross entropy and sigmoid to each element in vector
+            loss_calc.append(self.__cross_entropy(self.__sigmoid(x), y).numpy())
+
+        expected_loss = np.mean(loss_calc)
+
+        np.testing.assert_allclose(loss, expected_loss)
```
Comment (Blubberli) [inline tests/test_loss_functions.py:77]: This test fails on my computer because you're using double underscores which you usually would do to access a private variable. To access a static method you just use class.static_method, in this case the class = self, so self.sigmoid and self.cross_entropy work

File: scripts/loss_functions.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+def multi_class_cross_entropy(output, target):
+    """
+    combines log_softmax and nll_loss in a single function, is used for multiclass classification
+    :param output: the input to the loss function is the output as raw, unnormalized scores for each class,
+                    is of size (minibatch, C)
+    :param target: 1D tensor of size minibatch with range [0,C−1] for each value,
+    :return: loss: mean loss of all instances in batch
+    """
+    assert output.shape[0] == target.shape[0], "target shape is the number of batches in output"
+    assert target.dtype != 'float32', "target type has to be \"int\""  # maybe remove or use regex
```
Comment (Blubberli) [inline scripts/loss_functions.py]: I would specify the datatype in the docstrings and remove the assertion. if the user does not input parameters of the type specified in the docstrings he will get a type error which is fine. I think the shape assertion is more important and makes sense.

**Review by Blubberli — CHANGES_REQUESTED** [review]
Just a few things to change or clarify. I think it is very nice that you computed the loss manually even though of course we can expect the built in function to do the math correctly. 
One thing one could test is whether the loss is 0 if the logits are 1 for the correct target and 0 for all others and if the loss is > 0 if that is not the case. But such a test can also be added later, so it's up to you. 
Thanks for your engagement and precise coding work and documentation!

File: tests/test_loss_functions.py (hunk 2 above)
Comment (evvahu) [reply tests/test_loss_functions.py:77]: yeah, I forgot to change it back... because I initially tried to do cross_entropy and sigmoid in another helper static method and have cross_entropy and sigmoid as private methods. Probably doesn't make much sense and doesn't work anyway (--> private methods couldn't be called)

File: tests/test_loss_functions.py (hunk 1 above)
Comment (evvahu) [reply tests/test_loss_functions.py:19]: I honestly don't know why I did that...I must have copy-pasted it from someone :/

Comment (evvahu) [reply tests/test_loss_functions.py:19]: output_tensor has to be float though..right? because these are the "logits"

**Review by Blubberli — APPROVED** [review]
Very nice!
